## ให้นิสิตเขียนโปรแกรมเพื่อทำ sentiment analysis dataset Fastfood_Opinion.csv ด้วยเทคนิค Naïve Bayes และ Logistic Regression

In [ ]:
!pip install pythainlp

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.9/17.9 MB 26.4 MB/s eta 0:00:00


In [ ]:
!wget -O fastfood.data.csv 'https://docs.google.com/uc?export=download&id=1K0VgdwYagNidu5k_y5pnSrGrGIOBPsyS'

--2024-07-23 16:49:24--  https://docs.google.com/uc?export=download&id=1K0VgdwYagNidu5k_y5pnSrGrGIOBPsyS
Resolving docs.google.com (docs.google.com)... 64.233.180.100, 64.233.180.139, 64.233.180.113, ...
Connecting to docs.google.com (docs.google.com)|64.233.180.100|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://drive.usercontent.google.com/download?id=1K0VgdwYagNidu5k_y5pnSrGrGIOBPsyS&export=download [following]
--2024-07-23 16:49:24--  https://drive.usercontent.google.com/download?id=1K0VgdwYagNidu5k_y5pnSrGrGIOBPsyS&export=download
Resolving drive.usercontent.google.com (drive.usercontent.google.com)... 172.253.62.132, 2607:f8b0:4004:c19::84
Connecting to drive.usercontent.google.com (drive.usercontent.google.com)|172.253.62.132|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 348315 (340K) [application/octet-stream]
Saving to: ‘fastfood.data.csv’

fastfood.data.csv   100%[===================>] 340.15K  --.-KB/s   

In [ ]:
import pandas as pd
import numpy as np
from pythainlp.tokenize import word_tokenize
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.naive_bayes import MultinomialNB

In [ ]:
df = pd.read_csv('fastfood.data.csv')
df

class                                            message
0        0  11.30 น. ไปที่สาขาในปั๊มคาลเท็กซ์ ประชานุกูล ซ...
1        0  BK ทุกสาขาบริการดี โดยเฉพาะอย่างยิ่งเอมโพเรียม...
2        0  Burger King สาขา The Bright พระราม 2 บริการห่ว...
3        0  Burger King สาขาปั้ม ป.ต.ท.บ้านพร อ.ถลาง เมื่อ...
4        0  chicken strip ชิ้นเล็กมาก เทียบกับป้ายโฆษณาบนโต๊ะ
..     ...                                                ...
637      1                          ช่วงโปรฯ คุณภาพเกินราคาคะ
638      1                           อร่อยมาก ราคาโอเครเลยค่ะ
639      1                                    อาหารอร่อยมากคะ
640      1                      คุณภาพและการบริการระดับ 5 ดาว
641      1                                 อาหารอร่อยมากๆ ค่ะ

[642 rows x 2 columns]

### 1.กำหนดค่า feature ในรูปแบบต่างๆ และวิเคราะห์ผลลัพธ์ที่ได้

- Bag of Words Term Frequency

- Binary Term Frequency

- Bigram Term Frequency

## **Bag of Words Term Frequency**

In [ ]:
# ตัดคำใส่ Bag of Words
X = df['message'].astype(str) # feature
y = df['class'] # lable

X_tokenized = X.apply(word_tokenize, keep_whitespace=False)

vectorizer_bow = CountVectorizer(analyzer=lambda x: x)
X_bow = vectorizer_bow.fit_transform(X_tokenized)
print(X_bow.shape)


(642, 3542)


**Naive Bays**

In [ ]:
# Naive Bays
X_train, X_test, y_train, y_test = train_test_split(X_bow, y, test_size=0.2, random_state=8)

nb_classifier = MultinomialNB()
nb_classifier.fit(X_train, y_train)

y_pred = nb_classifier.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Naive Bays Accuracy:", accuracy)

print("Classification Report:")
print(classification_report(y_test, y_pred))


print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Naive Bays Accuracy: 0.937984496124031
Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.97      0.95        77
           1       0.96      0.88      0.92        52

    accuracy                           0.94       129
   macro avg       0.94      0.93      0.93       129
weighted avg       0.94      0.94      0.94       129

Confusion Matrix:
[[75  2]
 [ 6 46]]


In [ ]:
# ทดลองใส่ประโยคใหม่
new_text = ["ดีมากค่ะ แต่แพงขอไม่ซ้ำ", "ซื้อซ้ำแน่นอนค่ะ คุณภาพสูงเกิน", "แพงมาก เก็บตังไปซื้อข้าวดีกว่า", "อร่อยอยู่ค่ะ ไม่แย่", "ไม่ตรงปกเลย ได้เช็คของก่อนส่งมั้ยคะ"]
new_text = pd.Series(new_text)
new_text_tokenized = new_text.apply(word_tokenize, keep_whitespace=False)
new_text_bow = vectorizer_bow.transform(new_text_tokenized)

predicted_class = nb_classifier.predict(new_text_bow)
print("Predicted Class:", predicted_class) # 1 คือดี 0 คือแย่

# คำตอบที่ถูกคือ [0, 1, 0, 1, 0]

Predicted Class: [1 1 0 0 0]


**Logistic regression**

In [ ]:
# Logistic regression

X_train, X_test, y_train, y_test = train_test_split(X_bow, y, test_size=0.2, random_state=8)

lr_classifier = LogisticRegression()
lr_classifier.fit(X_train, y_train)

y_pred = lr_classifier.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Logistic Regression Accuracy:", accuracy)

print("Classification Report:")
print(classification_report(y_test, y_pred))


print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Logistic Regression Accuracy: 0.9147286821705426
Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.92      0.93        77
           1       0.89      0.90      0.90        52

    accuracy                           0.91       129
   macro avg       0.91      0.91      0.91       129
weighted avg       0.92      0.91      0.91       129

Confusion Matrix:
[[71  6]
 [ 5 47]]


In [ ]:
# ทดลองใส่ประโยคใหม่
new_text = ["ดีมากค่ะ แต่แพงขอไม่ซ้ำ", "ซื้อซ้ำแน่นอนค่ะ คุณภาพสูงเกิน", "แพงมาก เก็บตังไปซื้อข้าวดีกว่า", "อร่อยอยู่ค่ะ ไม่แย่", "ไม่ตรงปกเลย ได้เช็คของก่อนส่งมั้ยคะ"]
new_text = pd.Series(new_text)
new_text_tokenized = new_text.apply(word_tokenize, keep_whitespace=False)
new_text_bow = vectorizer_bow.transform(new_text_tokenized)

predicted_class = lr_classifier.predict(new_text_bow)
print("Predicted Class:", predicted_class) # 1 คือดี 0 คือแย่

# คำตอบที่ถูกคือ [0, 1, 0, 1, 0]

Predicted Class: [1 1 1 0 0]


## **Binary Term Frequency**

In [ ]:
# ตัดคำ Binary Term
X = df['message'].astype(str)
y = df['class']

vectorizer_bt = CountVectorizer(binary=True)
X_bt = vectorizer_bt.fit_transform(X)
print(X_bt.shape)


(642, 8183)


**Naive Bays**

In [ ]:
# Naive Bays
X_train, X_test, y_train, y_test = train_test_split(X_bt, y, test_size=0.2, random_state=8)

nb_classifier_bt = MultinomialNB()
nb_classifier_bt.fit(X_train, y_train)

y_pred = nb_classifier_bt.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Naive Bays use binary term Accuracy:", accuracy)

print("Classification Report:")
print(classification_report(y_test, y_pred))


print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Naive Bays use binary term Accuracy: 0.9147286821705426
Classification Report:
              precision    recall  f1-score   support

           0       0.92      0.94      0.93        77
           1       0.90      0.88      0.89        52

    accuracy                           0.91       129
   macro avg       0.91      0.91      0.91       129
weighted avg       0.91      0.91      0.91       129

Confusion Matrix:
[[72  5]
 [ 6 46]]


In [ ]:
# ทดลองใส่ประโยคใหม่
new_text = ["ดีมากค่ะ แต่แพงขอไม่ซ้ำ", "ซื้อซ้ำแน่นอนค่ะ คุณภาพสูงเกิน", "แพงมาก เก็บตังไปซื้อข้าวดีกว่า", "อร่อยอยู่ค่ะ ไม่แย่", "ไม่ตรงปกเลย ได้เช็คของก่อนส่งมั้ยคะ"]
new_text = pd.Series(new_text)
new_text_tokenized = new_text.apply(lambda x: ' '.join(word_tokenize(x, keep_whitespace=False)))
new_text_bt = vectorizer_bt.transform(new_text_tokenized)

predicted_class = nb_classifier_bt.predict(new_text_bt)
print("Predicted Class:", predicted_class) # 1 คือดี 0 คือแย่

# คำตอบที่ถูกคือ [0, 1, 0, 1, 0]

Predicted Class: [0 1 0 0 0]


**Logistic Regression**

In [ ]:
# Logistic Regression

X_train, X_test, y_train, y_test = train_test_split(X_bt, y, test_size=0.2, random_state=8)

lr_classifier_bt = LogisticRegression()
lr_classifier_bt.fit(X_train, y_train)

y_pred = lr_classifier_bt.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Logistic Regression use binary term Accuracy:", accuracy)

print("Classification Report:")
print(classification_report(y_test, y_pred))


print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Logistic Regression use binary term Accuracy: 0.875968992248062
Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.86      0.89        77
           1       0.81      0.90      0.85        52

    accuracy                           0.88       129
   macro avg       0.87      0.88      0.87       129
weighted avg       0.88      0.88      0.88       129

Confusion Matrix:
[[66 11]
 [ 5 47]]


In [ ]:
# ทดลองใส่ประโยคใหม่
new_text = ["ดีมากค่ะ แต่แพงขอไม่ซ้ำ", "ซื้อซ้ำแน่นอนค่ะ คุณภาพสูงเกิน", "แพงมาก เก็บตังไปซื้อข้าวดีกว่า", "อร่อยอยู่ค่ะ ไม่แย่", "ไม่ตรงปกเลย ได้เช็คของก่อนส่งมั้ยคะ"]
new_text = pd.Series(new_text)
new_text_tokenized = new_text.apply(lambda x: ' '.join(word_tokenize(x, keep_whitespace=False)))
new_text_bt = vectorizer_bt.transform(new_text_tokenized)

predicted_class = lr_classifier_bt.predict(new_text_bt)
print("Predicted Class:", predicted_class) # 1 คือดี 0 คือแย่

# คำตอบที่ถูกคือ [0, 1, 0, 1, 0]

Predicted Class: [0 1 1 0 0]


## **Bigram Term Frequency**

In [ ]:
# ตัดคำ Bigram Term
X = df['message'].astype(str)
y = df['class']

vectorizer_bg = CountVectorizer(ngram_range=(2, 2), max_features=2)
X_bg = vectorizer_bg.fit_transform(X)
print(X_bg.shape)

(642, 2)


**Naive Bayes**

In [ ]:
# Naive Bayes
X_train, X_test, y_train, y_test = train_test_split(X_bg, y, test_size=0.2, random_state=8)

nb_classifier_bg = MultinomialNB()
nb_classifier_bg.fit(X_train, y_train)

y_pred = nb_classifier_bg.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Naive Bays use bigram term Accuracy:", accuracy)

print("Classification Report:")
print(classification_report(y_test, y_pred))


print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Naive Bays use bigram term Accuracy: 0.6976744186046512
Classification Report:
              precision    recall  f1-score   support

           0       0.66      1.00      0.80        77
           1       1.00      0.25      0.40        52

    accuracy                           0.70       129
   macro avg       0.83      0.62      0.60       129
weighted avg       0.80      0.70      0.64       129

Confusion Matrix:
[[77  0]
 [39 13]]


In [ ]:
# ทดลองใส่ประโยคใหม่
new_text = ["ดีมากค่ะ แต่แพงขอไม่ซ้ำ", "ซื้อซ้ำแน่นอนค่ะ คุณภาพสูงเกิน", "แพงมาก เก็บตังไปซื้อข้าวดีกว่า", "อร่อยอยู่ค่ะ ไม่แย่", "ไม่ตรงปกเลย ได้เช็คของก่อนส่งมั้ยคะ"]
new_text = pd.Series(new_text)
new_text_tokenized = new_text.apply(lambda x: ' '.join(word_tokenize(x, keep_whitespace=False)))
new_text_bg = vectorizer_bg.transform(new_text_tokenized)

predicted_class = nb_classifier_bg.predict(new_text_bg)
print("Predicted Class:", predicted_class) # 1 คือดี 0 คือแย่

# คำตอบที่ถูกคือ [0, 1, 0, 1, 0]

Predicted Class: [0 0 0 0 0]


**Logistic Regression**

In [ ]:
# Logistic Regression
X_train, X_test, y_train, y_test = train_test_split(X_bg, y, test_size=0.2, random_state=8)

lr_classifier_bg = LogisticRegression()
lr_classifier_bg.fit(X_train, y_train)

y_pred = lr_classifier_bg.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Logistic Regression use binary term Accuracy:", accuracy)

print("Classification Report:")
print(classification_report(y_test, y_pred))


print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Logistic Regression use binary term Accuracy: 0.6976744186046512
Classification Report:
              precision    recall  f1-score   support

           0       0.66      1.00      0.80        77
           1       1.00      0.25      0.40        52

    accuracy                           0.70       129
   macro avg       0.83      0.62      0.60       129
weighted avg       0.80      0.70      0.64       129

Confusion Matrix:
[[77  0]
 [39 13]]


In [ ]:
# ทดลองใส่ประโยคใหม่
new_text = ["ดีมากค่ะ แต่แพงขอไม่ซ้ำ", "ซื้อซ้ำแน่นอนค่ะ คุณภาพสูงเกิน", "แพงมาก เก็บตังไปซื้อข้าวดีกว่า", "อร่อยอยู่ค่ะ ไม่แย่", "ไม่ตรงปกเลย ได้เช็คของก่อนส่งมั้ยคะ"]
new_text = pd.Series(new_text)
new_text_tokenized = new_text.apply(lambda x: ' '.join(word_tokenize(x, keep_whitespace=False)))
new_text_bg = vectorizer_bg.transform(new_text_tokenized)

predicted_class = lr_classifier_bg.predict(new_text_bg)
print("Predicted Class:", predicted_class) # 1 คือดี 0 คือแย่

# คำตอบที่ถูกคือ [0, 1, 0, 1, 0]

Predicted Class: [0 0 0 0 0]


คะแนนความถูกต้องที่ได้จากตัวเทรน (Accuracy Score) ได้ดังนี้

**Bag of Words Term Frequency**
- Naive Bayes 0.93
- Logistic Regression 0.91

**Binary Term Frequency**
- Naive Bayes 0.91
- Logistic Regression 0.87

**Bigram Term Frequency**
- Naive Bayes 0.69
- Logistic Regression 0.69


คะแนนความถูกต้องของประโยคที่เพิ่มใหม่ของแต่ละโมเดล ได้ดังนี้

**Bag of Words Term Frequency**
- Naive Bayes 3/5
- Logistic Regression 2/5

**Binary Term Frequency**
- Naive Bayes 4/5
- Logistic Regression 3/5

**Bigram Term Frequency**
- Naive Bayes 3/5
- Logistic Regression 3/5



> ตัว Bigram Term Frequency ไม่ว่าใช้กับโมเดลไหนก็ได้คะแนนเท่า ๆ กัน ไม่ต่างกันมาก



> ตัวโมเดล Naive Bayes ที่ใช้ Binary Term Frequency ได้คะแนนความถูกต้องข้อมูลเทรนใกล้เคียงกับโมเดล Logistic Regression ที่ใช้ Bag of Words Term Frequency แต่โมเดล Naive Bayes ที่ใช้ Binary Term Frequency ที่ทดลองด้วยประโยคใหม่ ได้คะแนนความถูกต้องมากกว่าโมเดล Logistic Regression ที่ใช้ Bag of Words Term Frequency

